File: frontend/src/lib/api/applications.ts
import { internalError, isRecord, postInit, request, toApiError } from '@/lib/api/http';
import type {
  Application,
  ApplicationList,
  ApplicationViewerRole,
  ApplyPayload,
  ApplyResult,
  ChatRoom,
} from '@/types/application';
import type { MaskedUser } from '@/types/application';

/**
 * 브라우저는 백엔드 주소를 직접 호출하지 않는다. 동일 출처 상대 경로로만 호출하고
 * next.config.ts의 rewrite가 FastAPI로 전달한다. 서버 컴포넌트에서는 baseUrl로
 * 절대 URL을 지정한다(설계서 11.2).
 */
const REQUESTS_BASE = '/api/requests';
const CHAT_ROOMS_BASE = '/api/chat-rooms';

const ALLOWED_FIELD_KEYS = ['offerPrice', 'message'] as const;

interface RequestOpts {
  signal?: AbortSignal;
  baseUrl?: string;
  cookie?: string;
}

function isMaskedUser(value: unknown): value is MaskedUser {
  return isRecord(value) && typeof value.id === 'string' && typeof value.maskedEmail === 'string';
}

function parseApplication(value: unknown): Application | null {
  if (!isRecord(value)) return null;

  const { id, requestId, seller, offerPrice, message, chatRoomId, createdAt } = value;
  if (
    typeof id !== 'string' ||
    typeof requestId !== 'string' ||
    !isMaskedUser(seller) ||
    typeof offerPrice !== 'number' ||
    typeof message !== 'string' ||
    typeof chatRoomId !== 'string' ||
    typeof createdAt !== 'string'
  ) {
    return null;
  }

  return { id, requestId, seller, offerPrice, message, chatRoomId, createdAt };
}

function isViewerRole(value: unknown): value is ApplicationViewerRole {
  return value === 'owner' || value === 'applicant' || value === 'member' || value === 'anonymous';
}

function parseApplicationList(value: unknown): ApplicationList | null {
  if (!isRecord(value)) return null;

  const { viewerRole, applicantCount, items } = value;
  if (!isViewerRole(viewerRole) || typeof applicantCount !== 'number' || !Array.isArray(items)) {
    return null;
  }

  const parsedItems = items.map(parseApplication);
  if (parsedItems.some((item) => item === null)) return null;

  return { viewerRole, applicantCount, items: parsedItems as Application[] };
}

function isChatRoomViewerRole(value: unknown): value is ChatRoom['viewerRole'] {
  return value === 'buyer' || value === 'seller';
}

function isRequestStatus(value: unknown): value is ChatRoom['request']['status'] {
  return value === 'open' || value === 'matched' || value === 'closed';
}

function parseChatRoom(value: unknown): ChatRoom | null {
  if (!isRecord(value)) return null;

  const {
    id,
    applicationId,
    viewerRole,
    request: requestSummary,
    buyer,
    seller,
    offerPrice,
    applicationMessage,
    createdAt,
  } = value;

  if (
    typeof id !== 'string' ||
    typeof applicationId !== 'string' ||
    !isChatRoomViewerRole(viewerRole) ||
    !isRecord(requestSummary) ||
    typeof requestSummary.id !== 'string' ||
    typeof requestSummary.title !== 'string' ||
    !isRequestStatus(requestSummary.status) ||
    typeof requestSummary.priceMin !== 'number' ||
    typeof requestSummary.priceMax !== 'number' ||
    !isMaskedUser(buyer) ||
    !isMaskedUser(seller) ||
    typeof offerPrice !== 'number' ||
    typeof applicationMessage !== 'string' ||
    typeof createdAt !== 'string'
  ) {
    return null;
  }

  return {
    id,
    applicationId,
    viewerRole,
    request: {
      id: requestSummary.id,
      title: requestSummary.title,
      status: requestSummary.status,
      priceMin: requestSummary.priceMin,
      priceMax: requestSummary.priceMax,
    },
    buyer,
    seller,
    offerPrice,
    applicationMessage,
    createdAt,
  };
}

export async function applyToRequest(
  requestId: string,
  payload: ApplyPayload,
  signal?: AbortSignal,
): Promise<ApplyResult> {
  const response = await request(`/${requestId}/applications`, postInit(payload, signal), REQUESTS_BASE);

  if (!response.ok) {
    const error = await toApiError(response, ALLOWED_FIELD_KEYS);
    // 지원 API는 409를 계약된 code(ALREADY_APPLIED/REQUEST_NOT_OPEN)로만 다룬다.
    // 본문이 깨져 공용 fallbackCode(409)가 EMAIL_ALREADY_EXISTS로 잘못 짐작하면
    // 인증 관련 오류로 오인하지 않도록 INTERNAL_ERROR로 바꾼다(설계서 11.2).
    if (response.status === 409 && error.code === 'EMAIL_ALREADY_EXISTS') {
      throw internalError(response.status);
    }
    throw error;
  }

  let body: unknown = null;
  try {
    body = await response.json();
  } catch {
    throw internalError(response.status);
  }

  if (!isRecord(body)) throw internalError(response.status);
  const application = parseApplication(body.application);
  const chatRoom = parseChatRoom(body.chatRoom);
  if (!application || !chatRoom) throw internalError(response.status);

  return { application, chatRoom };
}

export async function listApplications(
  requestId: string,
  opts: RequestOpts = {},
): Promise<ApplicationList> {
  const response = await request(
    `/${requestId}/applications`,
    { method: 'GET', signal: opts.signal, headers: opts.cookie ? { Cookie: opts.cookie } : undefined },
    `${opts.baseUrl ?? ''}${REQUESTS_BASE}`,
  );

  if (!response.ok) throw await toApiError(response, ALLOWED_FIELD_KEYS);

  let body: unknown = null;
  try {
    body = await response.json();
  } catch {
    throw internalError(response.status);
  }

  const list = parseApplicationList(body);
  if (!list) throw internalError(response.status);
  return list;
}

export async function getChatRoom(roomId: string, opts: RequestOpts = {}): Promise<ChatRoom> {
  const response = await request(
    `/${roomId}`,
    { method: 'GET', signal: opts.signal, headers: opts.cookie ? { Cookie: opts.cookie } : undefined },
    `${opts.baseUrl ?? ''}${CHAT_ROOMS_BASE}`,
  );

  if (!response.ok) throw await toApiError(response, []);

  let body: unknown = null;
  try {
    body = await response.json();
  } catch {
    throw internalError(response.status);
  }

  const chatRoom = parseChatRoom(body);
  if (!chatRoom) throw internalError(response.status);
  return chatRoom;
}

File: backend/app/services/applications.py
from __future__ import annotations

import uuid
from typing import Any
from uuid import UUID

from sqlalchemy import text
from sqlalchemy.exc import IntegrityError, OperationalError, SQLAlchemyError
from sqlalchemy.orm import Session

from app.core.config import Settings
from app.core.db_logging import log_database_failure
from app.db.models import PurchaseRequest, User
from app.repositories import applications as application_repository
from app.repositories import chat_rooms as chat_room_repository
from app.schemas.applications import ApplicationCreate, ApplyResponse
from app.services.errors import ServiceUnavailable
from app.services.requests import RequestNotFound, mask_email


class SelfApplicationForbidden(Exception):
    pass


class RequestNotOpen(Exception):
    pass


class AlreadyApplied(Exception):
    pass


class ChatRoomNotFound(Exception):
    pass


def _masked_user(user_id: UUID, email: str) -> dict[str, Any]:
    return {"id": user_id, "masked_email": mask_email(email)}


def _application_data(application, room_id: UUID, seller_email: str) -> dict[str, Any]:
    return {
        "id": application.id,
        "request_id": application.request_id,
        "seller": _masked_user(application.seller_id, seller_email),
        "offer_price": application.offer_price,
        "message": application.message,
        "chat_room_id": room_id,
        "created_at": application.created_at,
    }


def _room_data(
    *,
    room_id: UUID,
    application_id: UUID,
    room_created_at,
    request: PurchaseRequest,
    buyer_email: str,
    seller_id: UUID,
    seller_email: str,
    offer_price: int,
    message: str,
    viewer_role: str,
) -> ApplyResponse:
    return {
        "id": room_id,
        "application_id": application_id,
        "viewer_role": viewer_role,
        "request": {
            "id": request.id,
            "title": request.title,
            "status": request.status,
            "price_min": request.price_min,
            "price_max": request.price_max,
        },
        "buyer": _masked_user(request.buyer_id, buyer_email),
        "seller": _masked_user(seller_id, seller_email),
        "offer_price": offer_price,
        "application_message": message,
        "created_at": room_created_at,
    }


def _constraint_name(exc: IntegrityError) -> str | None:
    return getattr(getattr(exc, "orig", None), "diag", None) and getattr(exc.orig.diag, "constraint_name", None)


def apply_to_request(
    db: Session,
    *,
    request_id: UUID,
    seller: User,
    payload: ApplicationCreate,
    settings: Settings,
) -> dict[str, Any]:
    """Create one application and its private room in a single commit."""
    try:
        db.execute(
            text("SELECT set_config('lock_timeout', :timeout, true)"),
            {"timeout": f"{settings.application_lock_timeout_ms}ms"},
        )
        locked = application_repository.lock_request_for_application(db, request_id)
        if locked is None:
            db.rollback()
            raise RequestNotFound
        request, buyer_email = locked
        if request.buyer_id == seller.id:
            db.rollback()
            raise SelfApplicationForbidden
        if request.status != "open":
            db.rollback()
            raise RequestNotOpen

        application_id = uuid.uuid4()
        inserted = application_repository.insert_application(
            db,
            application_id=application_id,
            request_id=request.id,
            buyer_id=request.buyer_id,
            seller_id=seller.id,
            offer_price=payload.offer_price,
            message=payload.message,
        )
        if inserted is None:
            db.rollback()
            raise AlreadyApplied
        application_id, application_created_at = inserted
        room_id, room_created_at = chat_room_repository.create_for_application(
            db, room_id=uuid.uuid4(), application_id=application_id
        )

        # Validate the completed DTO before the sole commit.  If response
        # construction ever becomes invalid, neither INSERT is durable.
        application_data = {
            "id": application_id,
            "request_id": request.id,
            "seller": _masked_user(seller.id, seller.email),
            "offer_price": payload.offer_price,
            "message": payload.message,
            "chat_room_id": room_id,
            "created_at": application_created_at,
        }
        room_data = _room_data(
            room_id=room_id,
            application_id=application_id,
            room_created_at=room_created_at,
            request=request,
            buyer_email=buyer_email,
            seller_id=seller.id,
            seller_email=seller.email,
            offer_price=payload.offer_price,
            message=payload.message,
            viewer_role="seller",
        )
        response = ApplyResponse.model_validate({"application": application_data, "chat_room": room_data})
        db.commit()
        return response
    except IntegrityError as exc:
        db.rollback()
        constraint = _constraint_name(exc)
        if constraint == "uq_seller_applications_request_seller":
            raise AlreadyApplied from None
        if constraint == "ck_seller_applications_not_self":
            raise SelfApplicationForbidden from None
        log_database_failure("apply_to_request", exc, settings)
        raise ServiceUnavailable from None
    except (OperationalError, SQLAlchemyError) as exc:
        db.rollback()
        log_database_failure("apply_to_request", exc, settings)
        raise ServiceUnavailable from None
    except Exception:
        # Non-database failures raised before a successful commit (including
        # DTO validation) must not leave the application or room pending.
        db.rollback()
        raise


def list_applications(
    db: Session, *, request_id: UUID, viewer_id: UUID | None, settings: Settings
) -> dict[str, Any]:
    try:
        buyer_id = application_repository.request_exists_and_buyer(db, request_id)
        if buyer_id is None:
            raise RequestNotFound
        count = application_repository.count_for_request(db, request_id)
        if viewer_id is None:
            role = "anonymous"
        elif viewer_id == buyer_id:
            role = "owner"
        else:
            role = "applicant" if db.scalar(
                text(
                    "SELECT EXISTS (SELECT 1 FROM app_private.seller_applications "
                    "WHERE request_id = :request_id AND seller_id = :seller_id)"
                ),
                {"request_id": request_id, "seller_id": viewer_id},
            ) else "member"
        rows = application_repository.list_for_request(db, request_id=request_id, viewer_id=viewer_id, role=role)
        return {
            "viewer_role": role,
            "applicant_count": count,
            "items": [_application_data(application, room_id, seller_email) for application, room_id, seller_email in rows],
        }
    except RequestNotFound:
        raise
    except SQLAlchemyError as exc:
        db.rollback()
        log_database_failure("list_applications", exc, settings)
        raise ServiceUnavailable from None


def get_chat_room(
    db: Session, *, room_id: UUID, viewer: User, settings: Settings
) -> dict[str, Any]:
    try:
        record = chat_room_repository.get_for_participant(db, room_id=room_id, viewer_id=viewer.id)
        if record is None:
            raise ChatRoomNotFound
        room, application, request, buyer_email, seller_email = record
        return _room_data(
            room_id=room.id,
            application_id=application.id,
            room_created_at=room.created_at,
            request=request,
            buyer_email=buyer_email,
            seller_id=application.seller_id,
            seller_email=seller_email,
            offer_price=application.offer_price,
            message=application.message,
            viewer_role="buyer" if viewer.id == application.buyer_id else "seller",
        )
    except ChatRoomNotFound:
        raise
    except SQLAlchemyError as exc:
        db.rollback()
        log_database_failure("get_chat_room", exc, settings)
        raise ServiceUnavailable from None
